designs/2023-rule-options-defaults/README.md

```diff
@@ -0,0 +1,295 @@
+
+- The core ESLint docs: automating and therefore standardizing how rule docs pages phrase default values
+- typescript-eslint's docs: the same as the core ESLint docs
+- [`eslint-doc-generator`](https://github.com/bmish/eslint-doc-generator)
+
+We'll also want to mention this in the ESLint core custom rule documentation.
+
+## Drawbacks
+
+- This increases both conceptual and implementation complexity around rule options
+- As a future breaking change, this may break user rules - especially private rules we don't have visibility to
+
+## Backwards Compatibility Analysis
+
+The new `useRecursiveOptionDefaults` option being opt-in means this would not be a breaking change in the current version of ESLint.
+This RFC's intent is for that option to gradually be removed over time:
+
+1. ESLint v8: opt-in (default to `false`)
+2. ESLint v9: opt-out (default to `true`)
+3. ESLint v10: remove both `useRecursiveOptionDefaults` and `context.optionsRaw`
+
+Changing `useRecursiveOptionDefaults` to opt-out would be a breaking change as it impacts what values are passed to the `create` method of rules.
+That breaking change will impact rules which rely on the exact length or existing of properties in `context.options`.
+Rules that wish to preserve the legacy options behavior can switch to referring to the new `context.optionsRaw` property.
+
+In the draft PR containing this change, only 3 test files -rule tests for `indent`, `max-len`, and `no-multiple-empty-lines`- failed during `npm run test`.
+All were fixed with a direct find-and-replace from `context.options` to `context.optionsRaw` in those rules' implementations.
+
+## Alternatives
+
+### Non-Recursive Object Defaults
+
+Ajv provides its own support for `default` values with a [`useDefaults` option](https://ajv.js.org/guide/modifying-data.html#assigning-defaults).
+It's used in ESLint core already, in [`lib/shared/ajv.js`](https://github.com/eslint/eslint/blob/22a558228ff98f478fa308c9ecde361acc4caf20/lib/shared/ajv.js#L21).
+Sticking Ajv's own defaults logic would bring two benefits:
+
+- Aligning ESLint schema defaulting with the standard logic used by Ajv
+- Avoiding a custom defaulting implementation in ESLint by using Ajv's instead
+
+However, [`useDefaults` requires explicit empty object for intermediate object entries to work for nested properties](https://github.com/ajv-validator/ajv/issues/1710).
+That means:
+
+- Schemas with optional objects containing properties with default need to provide an explicit `default: {}` for those objects.
+- If those properties are themselves optional objects containing properties, both the outer and inner objects need explicit `default` values.
+
+An alternative to this RFC could be to fill in those default values in ESLint core rules and plugins.
+See [this Runkit with simulated Ajv objects](https://runkit.com/joshuakgoldberg/650a826da5839400082514ff) for examples of how rules would need to provide those default values.
```

Comment:
I'm not sure if it's necessary to specify nested objects in defaults. In particular, would it work the same with `default: {}` instead of `default: { outer: { middle: {} } }`, and `default: {}`  instead of `default: { middle: {} }`?

Reply:
> `default: {}` instead of `default: { outer: { middle: {} } }`

Then if the provided rule config doesn't explicitly provide a `middle` property, the result won't have one.

```plaintext
✅ Nested: [] -> [{\"outer\":{}}]
✅ Nested: [{}] -> [{}]
✅ Nested: [{\"outer\":{}}] -> [{\"outer\":{}}]
✅ Nested: [{\"outer\":{\"middle\":{}}}] -> [{\"outer\":{\"middle\":{\"inner\":true}}}]
✅ Nested: [{\"outer\":{\"middle\":{\"inner\":false}}}] -> [{\"outer\":{\"middle\":{\"inner\":false}}}]
```

> `default: {}` instead of `default: { middle: {} }`

Similarly, if the config doesn't explicitly specify a `middle`, even if it does provide an `outer`, the result won't have a `middle`.

With `default: { outer: { middle: {} } }` in the outer layer:

```plaintext
✅ Nested: [] -> [{\"outer\":{\"middle\":{\"inner\":true}}}]
✅ Nested: [{}] -> [{\"outer\":{}}]
✅ Nested: [{\"outer\":{}}] -> [{\"outer\":{}}]
✅ Nested: [{\"outer\":{\"middle\":{}}}] -> [{\"outer\":{\"middle\":{\"inner\":true}}}]
✅ Nested: [{\"outer\":{\"middle\":{\"inner\":false}}}] -> [{\"outer\":{\"middle\":{\"inner\":false}}}]
```

With `default: {}` in the outer layer:

```plaintext
✅ Nested: [] -> [{\"outer\":{}}]
✅ Nested: [{}] -> [{\"outer\":{}}]
✅ Nested: [{\"outer\":{}}] -> [{\"outer\":{}}]
✅ Nested: [{\"outer\":{\"middle\":{}}}] -> [{\"outer\":{\"middle\":{\"inner\":true}}}]
✅ Nested: [{\"outer\":{\"middle\":{\"inner\":false}}}] -> [{\"outer\":{\"middle\":{\"inner\":false}}}]
```